ethcore/engines/instant-seal/src/lib.rs

```diff
@@ -80,10 +78,10 @@ impl Engine for InstantSeal {
 		if block.transactions.is_empty() {
 			Seal::None
 		} else {
-			let mut sealed_blocks = self.sealed_blocks.lock();
 			let block_number = block.header.number();
-			if !sealed_blocks.contains(&block_number) {
-				sealed_blocks.insert(block_number);
+			let last_sealed_block = self.last_sealed_block.load(Ordering::SeqCst);
+			if block_number > last_sealed_block {
+				self.last_sealed_block.store(block_number, Ordering::SeqCst);
```

Comment:
We have a race condition here. We should do `compare_and_swap` and check if we were the first ones to update the block - if not we should return `Seal::None`.

Reply:
Yeah, it's too bad there isn't `fetch_max` in stable yet... Wouldn't it make more sense to use a Mutex/RwLock then?

Reply:
Actually `compare_and_swap` is just fine - we just don't create a new block if we know that some other thread already did. `fetch_max` would just make it in one atomic call vs two (if it actually translates to an atomic call), but it's negligible.